Issue43219: Move "System Default Domain" to Authentication page

RosalineP opened the pull request:
#### Rationale
The second property on the "Site Settings" page is called "System default domain (default domain for user log in)". This is really an authentication property and should be on the Authentication configuration page. Moving it there would make it more discoverable.

#### Related Pull Requests
* https://github.com/LabKey/ldap/pull/51
* https://github.com/LabKey/itn-tslk-modules/pull/11

#### Changes
* Refactor calls of `AppProps.getDefaultDomain()` to use `AuthenticationManager.getDefaultDomain()`
* Create `AuthenticationManager.getDefaultDomain()`
* Refactor calls of `WriteableAppProps.setDefaultDomain()` to `AuthenticationManager.setDefaultDomain()`
* Create `AuthenticationManager.setDefaultDomain()`
* Add deferred Java upgrade code that reads the old property value and saves it to the new property set
* Update front-end

RosalineP committed:
Initial rough

RosalineP committed:
Linting

RosalineP committed:
Merge branch 'develop' into fb_Issue43219

RosalineP committed:
Merge branch 'develop' into fb_Issue43219

RosalineP committed:
Final tweaks; add deferred java upgrade code

RosalineP committed:
Adjust css

RosalineP commented on api/src/org/labkey/api/security/AuthenticationManager.java:
```diff
@@ -257,19 +272,27 @@ public static void saveAuthSetting(User user, String key, boolean value)
             changed.forEach((k, v)->{
-                props.put(k, Boolean.toString(v));
-                addAuthSettingAuditEvent(user, k, v ? "enabled" : "disabled");
+                boolean vIsBoolean = v instanceof Boolean;
```
Not very neat but my shot at wrapping the save into saveAuthSettings(). Alternatively, I could pull out saving the string into a separate function?

labkey-adam replied:
Eh, I wouldn't worry about it... seems fine to me

RosalineP commented on api/src/org/labkey/api/settings/AppPropsImpl.java:
```diff
@@ -626,7 +620,6 @@ public static void populateSiteSettingsWithStartupProps()
                 LOG.debug("Setting site settings config property '" + prop.getName() + "' to '" + prop.getValue() + "'");
                 switch (prop.getName())
                 {
-                    case DEFAULT_DOMAIN_PROP -> writeable.setDefaultDomain(prop.getValue());
```
Is this still needed in this location, actually, since it's still a 'site setting' that gets populated here?

labkey-adam replied:
Good catch. Our syseng team is likely setting this property on hosted instances, so we need to coordinate with them. We need to email them know that, as of 21.7, there's a new way to set the default domain via startup properties (and we should verify that it works). It would be nice if this code would remain backward compatible for a while... it should accept the old property name but call the new AuthenticationManager setter AND log a warning stating its going away and listing the correct approach.

RosalineP replied:
Done!

RosalineP committed:
Remove comment to self

RosalineP commented on core/src/client/components/GlobalSettings.tsx:
```diff
@@ -22,17 +24,52 @@ const ROW_TEXTS = [
+                System Default Domain
+                <LabelHelpTip title="Tip">
+                    <div>
+                        <div> Default domain for user log in.</div>
```
Note to Adam: Please check in the UI that the tooltip helptext looks good to you

labkey-adam replied:
```suggestion
                        <div> Default domain for user sign in.</div>
```

RosalineP commented on core/src/org/labkey/core/CoreUpgradeCode.java:
```diff
@@ -201,4 +206,49 @@ public void migrateSecondaryAuthenticationConfigurations(ModuleContext context)
+        }
+    }
+
+    public Map<String, String> getProperties(Container c)
```
I aggressively moved all necessary calls to help functions into the method—let me know if this is not what you meant when we talked about this!

labkey-adam replied:
Since this is a `@DeferredUpgrade`, it's safe to call into `AppPropsImpl`... but you'd have to make lookupStringValue() `public`. Oh, and AppPropsImpl isn't public. Here's my suggestion (not tested, but should work):

```
        String defaultDomain = (new AbstractWriteableSettingsGroup(){
            @Override
            protected String getGroupName()
            {
                return "site settings";
            }

            @Override
            protected String getType()
            {
                return SITE_CONFIG_NAME;
            }

            private String getDefaultDomain()
            {
                return lookupStringValue(DEFAULT_DOMAIN_PROP, "");
            }
        }).getDefaultDomain();
```

RosalineP replied:
Done!

RosalineP commented:
@labkey-adam 
I checked around and it looks like the other core executeJavaUpgradeCodes don't bump a value in `getSchemaVersion()` within `CoreModule` as is done with general upgrade scripts for other modules. Is the core module a special case?

**Edit:**
Nevermind, Josh cleared this up for me by pointing out the `SchemaVersion` setting in `core/module.properties`!

RosalineP committed:
Bump version in module.properties for core

labkey-alan commented on core/src/client/AuthenticationConfiguration/AuthenticationConfiguration.tsx:
```diff
@@ -95,21 +101,20 @@ export class App extends PureComponent<{}, Partial<State>> {
-                dirty: !isEquivalent(newState, globalSettingsDirtinessData)
-            })
-        );
+        const newState = { ...this.state.globalSettings, [id]: value };
```
This should be computed inside the this.setState call below and should reference `prevState` instead of `this.state` when doing so, that way you are guaranteed to be calculating state on the most up to date state.

RosalineP replied:
Oops, thanks for the catch. Done!

labkey-alan reviewed (approved):
From a frontend perspective it looks good, though there is a setState call that should probably be changed to compute state via the state object passed to setState just to be safe.

RosalineP committed:
Merge branch 'develop' into fb_Issue43219

RosalineP committed:
Apply Alan's feedback

Commit:
Item #8853: Ontology, Form insert and update (#2295)

All using concept picker for ontology columns on the row Insert/update pages for most domain types (when ontology module installed)

RosalineP committed:
Issue41645: Add User button on Project Users Banner (#2337)

labkey-jeckels committed:
Issue 43342: Use case-insensitive sort consistently in file browser (#2341)

Co-authored-by: labkey-jeckels <Josh Eckels>

labkey-alan committed:
Update @labkey/build to 4.0.0 (#2334)

* Fix paths in experimentRunGraphView.jsp and createPipelineTrigger.jsp
* Fix clean command for pipeline package.json

RosalineP committed:
Issue43330: Sample Type delete message displays names in random order (#2340)

labkey-klum committed:
Fb assay file watcher (#2311)

cnathe committed:
Item 8958: core-components page updates for SM admin page component changes (#2331)

* Update @labkey/components package version
* core-components updates for PermissionAssignments and SiteUsersGridPanel component changes

XingY committed:
Item 8908: Enable sample actions from Source sample grid

labkey-adam committed:
Standard support for "with timezone" JDBC types (#2347)

cnathe committed:
Issue 43334: Check for LABKEY.moduleContext to prevent JS error if session invalid check called when the moduleContext isn't available (#2344)

RosalineP committed:
SampleTypeLinkToStudyTest: Add nullcheck to prevent crawler error (#2345)

RosalineP committed:
Fix S3ImportTTestt test outage (#2336)

Commit:
Add the concept code w/o the ontology prefix to the identifiers for search. (#2258)

labkey-jeckels committed:
Issue 43368: Infinite redirect when hitting assay upload wizard (#2350)

labkey-kevink committed:
per-user mini-profiler settings and start minimized (#2349)

labkey-susanh committed:
Item 9012: Update application navigation menu behavior (#2343)

labkey-adam committed:
Move more specimen code (#2326)

labkey-matthewb committed:
Remove hard-coded Ext3 dependency in DataView (#2342)

* remove ext3 dependency in DataView
* <labkey:loadClientDependencies>

labkey-klum committed:
43373: Link to study : Assay's don't support fields with Visit concept URIs (#2354)

labkey-adam commented on core/src/org/labkey/core/admin/customizeSite.jsp:
```diff
@@ -177,13 +177,9 @@ Click the Save button at any time to accept the current settings and continue.</
 
 <tr>
-    <td colspan=2>Set default domain for user sign in and base server url (<%=bean.helpLink%>)</td>
+    <td colspan=2>Set base server url (<%=bean.helpLink%>)</td>
```
```suggestion
    <td colspan=2>URL settings (<%=bean.helpLink%>)</td>

```

labkey-klum committed:
43366: Users without AddUserPermission can still create new users from the manage group UI (#2348)

RosalineP committed:
Wording update from Adam

Co-authored-by: Adam Rauch <adam@labkey.com>

RosalineP committed:
Wording updates from Adam, the second

Co-authored-by: Adam Rauch <adam@labkey.com>

RosalineP committed:
Merge branch 'develop' into fb_Issue43219

# Conflicts:
#	experiment/src/org/labkey/experiment/ExperimentModule.java
#	specimen/src/org/labkey/specimen/security/roles/SpecimenCoordinatorRole.java

RosalineP committed:
Implement feedback

RosalineP commented:
Misclicked and accidentally re-requested from Alan; feel free to ignore that.

labkey-adam commented on api/src/org/labkey/api/settings/AppPropsImpl.java:
```diff
@@ -626,7 +621,10 @@ public static void populateSiteSettingsWithStartupProps()
-                    case DEFAULT_DOMAIN_PROP -> writeable.setDefaultDomain(prop.getValue());
+                    case DEFAULT_DOMAIN_PROP -> {
+                        AuthenticationManager.setDefaultDomain(prop.getValue());
+                        LOG.warn("DEFAULT_DOMAIN_PROP here will be deprecated in 21.11 in favor of using DEFAULT_DOMAIN in AuthenticationManager.populateSettingsWithStartupProps()");
```
```suggestion
                        LOG.warn("Support for the \"SiteSettings.defaultDomain\" startup property will be removed shortly; use \"Authentication.DefaultDomain\" instead.");

```

labkey-adam commented on core/src/org/labkey/core/login/LoginController.java:
```diff
@@ -1931,7 +1932,7 @@ public boolean handlePost(SetPasswordForm form, BindException errors) throws Exc
                         {
                             String defaultDomain = userEmailAddress.substring(atSign + 1);
                             WriteableAppProps appProps = AppProps.getWriteableInstance();
```
No need to touch appProps... remove this line

RosalineP replied:
Done!

labkey-adam commented on core/src/org/labkey/core/login/LoginController.java:
```diff
@@ -1931,7 +1932,7 @@ public boolean handlePost(SetPasswordForm form, BindException errors) throws Exc
                             WriteableAppProps appProps = AppProps.getWriteableInstance();
-                            appProps.setDefaultDomain(defaultDomain);
+                            AuthenticationManager.setDefaultDomain(defaultDomain);
                             appProps.save(null);
```
...and remove this line

RosalineP replied:
Done!

labkey-adam reviewed (changes requested):
A bit of cleanup...

labkey-adam committed:
Simplify and reuse for consistency

RosalineP committed:
Update wording

Co-authored-by: Adam Rauch <adam@labkey.com>

RosalineP committed:
Merge branch 'develop' into fb_Issue43219

RosalineP committed:
Apply feedback

RosalineP committed:
Update syntax in sqlserver script

RosalineP committed:
Merge branch 'develop' into fb_Issue43219
